plugins/nf-quilt/src/test/nextflow/quilt/QuiltProductTest.groovy

```diff
@@ -141,15 +141,37 @@ class QuiltProductTest extends QuiltSpecification {
         Map bad_meta = meta + ['Type': 'Workflow']
         Map skip_meta = ['metadata': 'SKIP']
 
+        when:
+        makeWriteProduct() // no metadata
+        then:
+        thrown(com.quiltdata.quiltcore.workflows.WorkflowException)
+
         expect:
-        makeWriteProduct().pubStatus == 1 // no metadata
-        makeWriteProduct(meta).pubStatus == 0 // valid metadata
-        makeWriteProduct().pubStatus == 1 // invalid default metadata
-        makeWriteProduct(bad_meta).pubStatus == 1 // invalid explicit metadata
-        makeWriteProduct(skip_meta).pubStatus == 1 // no default metadata
+        makeWriteProduct(meta) // valid metadata
+
+        when:
+        makeWriteProduct() // invalid default metadata
+        then:
+        thrown(com.quiltdata.quiltcore.workflows.WorkflowException)
+
+        when:
+        makeWriteProduct(bad_meta) // invalid explicit metadata
+        then:
+        thrown(com.quiltdata.quiltcore.workflows.WorkflowException)
+
+        when:
+        makeWriteProduct(skip_meta) // no default metadata
+        then:
+        thrown(com.quiltdata.quiltcore.workflows.WorkflowException)
+
         // NOTE: push does NOT update local registry
+        expect:
         makeWriteProduct().pkg.install() // try to merge existing metadata
-        makeWriteProduct(skip_meta).pubStatus == 1 // still fails on implicit metadata
+
+        when:
+        makeWriteProduct(skip_meta) // still fails on implicit metadata
+        then:
+        thrown(com.quiltdata.quiltcore.workflows.WorkflowException)
```

Comment:
I see you already defined one here.

quiltcore.package.PushException?

Reply:
I actually never created our own `PushException`, but just used `java.io.IOException`. Still don't know what's best.

Reply:
java.io.IOException is fine. The user shouldn't really have control over the Session.